### PR TITLE
Add  person re-identification module

### PR DESCRIPTION
**Pull request by xiaolai-sqlai**
Add market1501.py to download, decompress and make label automatically.
Add a strong baseline with resnet50 in market1501. The maximum is 93.1/80.6 rank/mAP, and the average is 92.5/80.1 rank/mAP, ,which is far higher than other baseline this year with only 88/71 rank1/mAP.

### CODE REVIEW

**Comment by hetong007** [comment]
PR from the Re-ID summer code project. @zhreshold @zhanghang1989 feel free to review and comment.

**Comment by mli** [comment]
Job PR-298-1 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/1/index.html 
Code coverage of this PR:  vs. Master:

File: scripts/datasets/market1501.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    with open(osp.join(exdir, 'train.txt'), 'w') as f:
+        for i, key in enumerate(train_list):
+            for item in train_list[key]:
+                f.write(item['name']+" "+str(i)+" "+item["pcam"]+"\n")
+    print("Make Label List Done")
+
+
+def main():
+    name = "Market-1501-v15.09.15"
+    url = "http://188.138.127.15:81/Datasets/"+name+".zip"
```
Comment (zhreshold) [inline scripts/datasets/market1501.py]: is this a private server? I cannot download the file. We'd better host it on s3 if the license permits

File: scripts/re-id/baseline/test.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    good_index = np.setdiff1d(query_index, camera_index, assume_unique=True)
+    junk_index1 = np.argwhere(gl==-1)
+    junk_index2 = np.intersect1d(query_index, camera_index)
+    junk_index = np.append(junk_index2, junk_index1) #.flatten())
+    
+    CMC_tmp = compute_mAP(index, good_index, junk_index)
+    return CMC_tmp
+
+
+def compute_mAP(index, good_index, junk_index):
```
Comment (hetong007) [inline scripts/re-id/baseline/test.py]: It's good to have one under https://github.com/dmlc/gluon-cv/tree/master/gluoncv/utils/metrics

File: scripts/re-id/baseline/process/data_read.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from __future__ import absolute_import
+import os, numbers
+
+from mxnet.gluon.data import dataset
+from mxnet import image, nd
+
+
+class ImageTxtDataset(dataset.Dataset):
```
Comment (hetong007) [inline scripts/re-id/baseline/process/data_read.py]: Please move to https://github.com/dmlc/gluon-cv/tree/master/gluoncv/data

File: scripts/re-id/baseline/process/transforms.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
+from __future__ import absolute_import
+import os, numbers, random, math
+
+from mxnet.gluon.data import dataset
+from mxnet import image, nd
+from mxnet.gluon import Block
+
+
+class Pad(Block):
```
Comment (hetong007) [inline scripts/re-id/baseline/process/transforms.py]: transform can be moved to https://github.com/dmlc/gluon-cv/blob/master/gluoncv/data/transforms/block.py

File: scripts/re-id/baseline/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
+## Reproducing market1501 experiments
+    1.Unlabeled Samples Generated by GAN Improve the Person Re-identification Baseline in vitro(iccv 2017)
```
Comment (hetong007) [inline scripts/re-id/baseline/README.md]: Link to arxiv is preferred.

File: scripts/re-id/baseline/README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+Considering the randomness for model training, these results may not be perfectly reproduced. A large number of experiments show that average result for our final model is 92.5 rank1 and 80.1 mAP, which is also get greatly improved compared with other results with resnet50 in Market1501.
+
+Base setting with single query: 256*128, epoch=25,50,75, batch size=32, init lr=3.5e-4, add batch norm after feature layer, average pooling, w/o activation
+
+result | rank1  | ran5 | rank10 | mAP
+--- | --- | --- | --- | ---
+baseline in 1 | 73.69 | - | - | 51.48
+baseline in 2 | 87.5 | 94.9 | 96.7 | 71.4
+paper 3 | 91.7 | - | - | 78.8
```
Comment (hetong007) [inline scripts/re-id/baseline/README.md]: Why is this name in another format comparing to the above two

**Comment by mli** [comment]
Job PR-298-3 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/3/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-4 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/4/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-5 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/5/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-6 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/6/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-8 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/8/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-10 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/10/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-298-24 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-298/24/index.html 
Code coverage of this PR:  vs. Master:

**Comment by hetong007** [comment]
@zhreshold @zhanghang1989 do you have any comments? Maybe we can have it in our 0.3 release.

**Comment by zhreshold** [comment]
Essentially no big issue to me, `data_read.py` and `read_label.py` are a bit weird, which could be merged into one file called reid.py inside `gluoncv/data/market1501/` for example

**Comment by hetong007** [comment]
I'm merging this as part of our 0.3 release. @xiaolai-sqlai great job!

**Pull request merged**